### PR TITLE
Add missing type name parsing test coverage

### PR DESCRIPTION
**Pull request by adamsitnik**
While working on the new Type Name Parser API (#97566) I've discovered some testing gaps.

Namely:

1. Reporting parse error for a backslash that was used to escape a character that is not legal to escape inside a type name: https://github.com/dotnet/runtime/blob/7185df8eda6c25d101170a2809094adc9f493b51/src/libraries/Common/src/System/Reflection/TypeNameParser.cs#L326-L331
2. Ignore leading '.' for type names without namespace:
https://github.com/dotnet/runtime/blob/7185df8eda6c25d101170a2809094adc9f493b51/src/libraries/Common/src/System/Reflection/TypeNameParser.cs#L200-L207
3. Supporting all kinds of leading whitespaces (existing test use only ` ` (space character)):
https://github.com/dotnet/runtime/blob/7185df8eda6c25d101170a2809094adc9f493b51/src/libraries/Common/src/System/Reflection/TypeNameParser.cs#L439-L450
4. Type names with escape characters in their name (illegal for C# and F#, but valid for IL)
https://github.com/dotnet/runtime/blob/7185df8eda6c25d101170a2809094adc9f493b51/src/libraries/Common/src/System/Reflection/TypeNameParser.cs#L628-L631
cc @GrabYourPitchforks @jeffhandley

### CODE REVIEW

**Comment by jkotas** [comment]
The tests are failing on Mono. Mono does not use the unified parser for all code paths yet - https://github.com/dotnet/runtime/issues/45033 . I think you can disable the failing tests on Mono against this issue.

**Comment by jkotas** [comment]
While you are on it, could you please also add a coverage for byref-of-byref and pointer-of-byref (e.g. here https://github.com/dotnet/runtime/blob/6d4fc1ad4725ca4cc77214b93bc1d37a6fa3553d/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs#L525)? We have been missing coverage for it - see #98426.

**Comment by adamsitnik** [comment]
@jkotas I've addressed your feedback, PTAL. Thanks!

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs [hunk 1]
```diff
@@ -513,6 +513,19 @@ public void GetTypeByName_ValidType_ReturnsExpected(string typeName, Type expect
             Assert.Equal(expectedType, Type.GetType(typeName.ToLower(), throwOnError: false, ignoreCase: true));
         }
 
+        public static IEnumerable<object[]> GetTypeByName_Invalid_ClrArguments()
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs:516]:
```suggestion
        public static IEnumerable<object[]> GetTypeByName_InvalidElementType()
```
I think this is a better description for this set of test cases.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs [hunk 2]
```diff
@@ -513,6 +513,19 @@ public void GetTypeByName_ValidType_ReturnsExpected(string typeName, Type expect
...
+        {
+            Type expectedException = PlatformDetection.IsMonoRuntime
+                ? typeof(ArgumentException) // https://github.com/dotnet/runtime/issues/45033
+                : typeof(TypeLoadException);
+
+            yield return new object[] { "System.Int32&&", expectedException, true };
+            yield return new object[] { "System.Int32&*", expectedException, true };
+            yield return new object[] { "System.Int32&[]", expectedException, true };
+            yield return new object[] { "System.Int32&[*]", expectedException, true };
+            yield return new object[] { "System.Int32&[,]", expectedException, true };
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs:526]:
```suggestion
            yield return new object[] { "System.Int32&[,]", expectedException, true };
            yield return new object[] { "System.Void[]", expectedException, true };
            yield return new object[] { "System.TypedReference[]", expectedException, true };
```
More interesting cases

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs [hunk 3]
```diff
@@ -522,6 +535,7 @@ public void GetTypeByName_ValidType_ReturnsExpected(string typeName, Type expect
         [InlineData("Outside`1[System.Boolean, System.Int32]", typeof(ArgumentException), true)]
         [InlineData(".System.Int32", typeof(TypeLoadException), false)]
         [InlineData("..Outside`1", typeof(TypeLoadException), false)]
+        [MemberData(nameof(GetTypeByName_Invalid_ClrArguments))]
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs:538]:
```suggestion
        [MemberData(nameof(GetTypeByName_InvalidElementType))]
```

**Review by jkotas — APPROVED** [review]
Thanks!

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs [hunk 4]
```diff
@@ -513,6 +513,21 @@ public void GetTypeByName_ValidType_ReturnsExpected(string typeName, Type expect
...
+                ? typeof(ArgumentException) // https://github.com/dotnet/runtime/issues/45033
+                : typeof(TypeLoadException);
+
+            yield return new object[] { "System.Int32&&", expectedException, true };
+            yield return new object[] { "System.Int32&*", expectedException, true };
+            yield return new object[] { "System.Int32&[]", expectedException, true };
+            yield return new object[] { "System.Int32&[*]", expectedException, true };
+            yield return new object[] { "System.Int32&[,]", expectedException, true };
+            yield return new object[] { "System.Void[]", expectedException, true };
+            yield return new object[] { "System.TypedReference[]", expectedException, true };
```
Comment (adamsitnik) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs:528]: @jkotas these two test cases are falling on Mono. Is my understanding correct that with https://github.com/dotnet/runtime/pull/94835 they should be working fine?

```suggestion

            if (!PlatformDetection.IsMonoRuntime)
            {
                yield return new object[] { "System.Void[]", expectedException, true };
                yield return new object[] { "System.TypedReference[]", expectedException, true };
            }
```

**Pull request merged**